src/contract/aci.ts

```diff
@@ -0,0 +1,516 @@
+    const onChanBytecode = (await getContractByteCode(contractAddress as unknown as EncodedData<'ct'>, { onNode })).bytecode
+    const isValid: boolean = source != null
+      ? await onCompiler.compilerApi.validateByteCode(
+        { bytecode: onChanBytecode, source, options: instance.options }
+      ).then(() => true, () => false)
+      : bytecode === onChanBytecode
+    if (!isValid) throw new BytecodeMismatchError(source != null ? 'source' : 'bytecode')
+  }
+
+  /**
+  * Compile contract
+  * @alias module:@aeternity/aepp-sdk/es/contract/aci
+  * @return bytecode
+  */
+  instance.compile = async (options = {}): Promise<string> => {
+    if (instance.bytecode != null) throw new IllegalArgumentError('Contract already compiled')
+    if (instance.source == null) throw new IllegalArgumentError('Can\'t compile without source code')
+    const { bytecode }: { bytecode: string } = await onCompiler.compilerApi.compileContract({
+      code: instance.source, options: { ...instance.options, ...options }
+    })
+    instance.bytecode = bytecode
+    return instance.bytecode
+  }
+
+  const sendAndProcess = async (tx: string, options: any): Promise<{
+    result?: ContractInstance['deployInfo']['result']
+    hash: string
+    tx: object
+    txData: TxData
+    rawTx: string
+  }> => {
+    const txData = await onAccount.send(tx, options) // TODO onAccount shouldn't have .send method
+    const result = {
+      hash: txData.hash, tx: unpackTx(txData.rawTx), txData, rawTx: txData.rawTx
+    }
+    if (txData.blockHeight == null) return result
+    const { callInfo } = await onCompiler.api.getTransactionInfoByHash(txData.hash)
+    Object.assign(result.txData, callInfo) // TODO: don't duplicate data in result
+    handleCallError(callInfo, tx)
+    return { ...result, result: callInfo }
+  }
+  const handleCallError = ({ returnType, returnValue }: {
+    returnType: ReturnType
+    returnValue: EncodedData<EncodingType> }, transaction: string): void => {
+    let message: string
+    switch (returnType) {
+      case 'ok': return
+      case 'revert':
+        message = instance.calldata.decodeFateString(returnValue)
+        break
+      case 'error':
+        message = decode(returnValue).toString()
+        break
+    }
+    throw new NodeInvocationError(message, transaction)
+  }
+
+  instance._estimateGas = async (
+    name: string, params: any[], options: object): Promise<number> => {
+    const { result: { gasUsed } } =
+      await instance.call(name, params, { ...options, callStatic: true })
+    // taken from https://github.com/aeternity/aepp-sdk-js/issues/1286#issuecomment-977814771
+    return Math.floor(gasUsed * 1.25)
+  }
+
+  /**
+  * Deploy contract
+  * @alias module:@aeternity/aepp-sdk/es/contract/aci
+  * @param params Contract init function arguments array
+  * @param options
+  * @returns deploy info
+  */
+  instance.deploy = async (params = [], options: object): Promise<ContractInstance['deployInfo']> => {
```

Comment:
options should combine options types of `instance.compile`, `instance.call`, `onAccount.address`, `sendAndProcess`